Libraries/LibWeb/Painting/ViewportPaintable.cpp

```diff
@@ -166,14 +166,29 @@ void ViewportPaintable::assign_clip_frames()
                 continue;
             }
             auto const& block_paintable_box = static_cast<PaintableBox const&>(*paintable);
-            auto block_overflow_x = block_paintable_box.computed_values().overflow_x();
-            auto block_overflow_y = block_paintable_box.computed_values().overflow_y();
-            if (block_overflow_x != CSS::Overflow::Visible && block_overflow_y != CSS::Overflow::Visible) {
-                auto rect = block_paintable_box.absolute_padding_box_rect();
-                clip_frame.add_clip_rect(rect, block_paintable_box.normalized_border_radii_data(ShrinkRadiiForBorders::Yes), block_paintable_box.enclosing_scroll_frame());
+            bool clip_x = paintable->computed_values().overflow_x() != CSS::Overflow::Visible;
+            bool clip_y = paintable->computed_values().overflow_y() != CSS::Overflow::Visible;
+
+            auto clip_rect = block_paintable_box.overflow_clip_edge_rect();
+            if (block_paintable_box.get_clip_rect().has_value()) {
+                clip_rect.intersect(block_paintable_box.get_clip_rect().value());
+                clip_x = true;
+                clip_y = true;
             }
-            if (auto css_clip_property_rect = block_paintable_box.get_clip_rect(); css_clip_property_rect.has_value()) {
-                clip_frame.add_clip_rect(css_clip_property_rect.value(), {}, block_paintable_box.enclosing_scroll_frame());
+
+            if (clip_x || clip_y) {
+                if (clip_x && clip_y) {
+                    clip_frame.add_clip_rect(clip_rect, block_paintable_box.normalized_border_radii_data(ShrinkRadiiForBorders::Yes), block_paintable_box.enclosing_scroll_frame());
+                } else {
+                    if (clip_x) {
+                        clip_rect.set_top(0);
+                        clip_rect.set_bottom(CSSPixels::max_integer_value);
+                    } else {
+                        clip_rect.set_left(0);
+                        clip_rect.set_right(CSSPixels::max_integer_value);
+                    }
+                    clip_frame.add_clip_rect(clip_rect, {}, block_paintable_box.enclosing_scroll_frame());
```

Comment:
is that correct to omit border radius application when clipping is only needed in one axis?